### PR TITLE
Upgrade to rust 1.52.1

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR depends on #9074.

This is part of upgrading from zexe to arkworks, which requires an upgrade to the latest rust (due to non-backwards-compatible typechecker changes in rust 1.49).

The changes in the `marlin` and `zexe` repos are minimal, and a build of this PR is able to connect and sync to mainnet, but this probably requires further testing to ensure that the new compiler doesn't cause issues with the existing OCaml-to-rust bindings.

Checklist:

- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them:

### CODE REVIEW

**Comment by aneesharaines** [comment]
@deepthiskumar  to help test this change

File: README-dev.md [hunk 1]
```diff
@@ -31,7 +31,7 @@ The short version:
 
 - Invoke `make macos-setup`
   - If this is your first time using OCaml, be sure to run `eval $(opam config env)`
-- Invoke `rustup default 1.45.2`
+- Invoke `rustup default 1.52.1`
```
Comment (mimoo) [inline README-dev.md:34]: you shouldn't need to do this as it's already specified in the `rust-toolchain.toml` file

Comment (mrmr1993) [reply README-dev.md:34]: There isn't a `rust-toolchain.toml` in the `marlin_plonk_bindings/stubs` library, so for now this is correct. I'll check in that file and update here.

**Comment by mrmr1993** [comment]
We've seen [a segfault](https://console.cloud.google.com/logs/query;cursorTimestamp=2021-07-01T09:00:54.356251844Z;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22o1labs-192920%22%0Aresource.labels.location%3D%22us-central1%22%0Aresource.labels.cluster_name%3D%22coda-infra-central1%22%0Aresource.labels.namespace_name%3D%22test-labels%22%0Aresource.labels.container_name%3D%22coda%22%0Aresource.labels.pod_name%3D%22whale-block-producer-2-58fb5cdbd6-2pcnz%22%0Atimestamp%3D%222021-07-01T09:00:54.356207863Z%22%0AinsertId%3D%221dy1as6g3vw7gmt%22;timeRange=2021-07-01T08:55:00.000Z%2F2021-07-01T09:01:30.000Z?project=o1labs-192920) on `test-labels` (the net running this PR) :sob: I've pulled in #8774 to get more info if it happens again, and merged compatible to test that CI still passes against the latest changes.

This is the first memory corruption issue in 9 days of network uptime, across 8 nodes. Given the infrequency, and given that we've also seen exactly one [segfault on devnet](https://console.cloud.google.com/logs/query;cursorTimestamp=2021-06-24T06:52:59.297051234Z;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22o1labs-192920%22%0Aresource.labels.location%3D%22us-east4%22%0Aresource.labels.cluster_name%3D%22coda-infra-east4%22%0Aresource.labels.namespace_name%3D%22devnet2%22%0Aresource.labels.container_name%3D%22coda%22%0Aseverity%3D%22ERROR%22%0Atimestamp%3D%222021-06-24T06:52:48.384385361Z%22%0AinsertId%3D%221s3fg3kg1sn4f6q%22;timeRange=2021-06-23T11:26:23.564Z%2F2021-07-01T11:26:23.564436Z?project=o1labs-192920) within the same period, I would prefer not to block merging this on discovering the source of that fault, especially since it may not be the cause at all.

**Review by lk86 — APPROVED** [review]

**Pull request merged**